File: 2015/17/rust/src/main.rs
use permutator::LargeCombinationIterator;
use std::collections::HashMap;
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const LITERS: usize = 150;

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut list: Vec<usize> = Vec::new();
    let mut count = 0;
    let mut map: HashMap<usize, usize> = HashMap::new();

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
        list.push(line.parse()?);
    }

    // --- Combinations ---
    for i in 1 ..= list.len() {
        let combinator = LargeCombinationIterator::new(&list, i);
        for c in combinator {
            let total: usize = c.iter().copied().sum();
            if total == LITERS {
                count += 1;
                *map.entry(c.len()).or_insert(0) += 1;
            }
        }
    }

    // --- Puzzle 1 ---
    println!("1. Combinations of containers: {}", count);

    // --- Puzzle 2 ---
    let min = map.iter().map(|(k, _)| k).min().unwrap();
    println!("2. Number of different ways: {}", map[min]);

    Ok(())
}

File: 2015/17/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";
import combinationN from "../../../commons/javascript/combinator.mjs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const LITERS = 150

// --- Variables ---
let list = [];
let count = 0;
let map = {};

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	list.push((Number(line)));
});
await new Promise(res => reader.once("close", res));

// --- Combinations ---
for (let i = 1; i <= list.length; i++) {
	for (const c of combinationN(list, i)) {
		let total = c.reduce((acc, x) => acc + x);
		if (total == LITERS) {
			count++;
			const len = c.length;
			if (!map.hasOwnProperty(len)) map[len] = 0;
			map[len]++;
		}
	}
}

// --- Puzzle 1 ---
console.log("1. Combinations of containers: ", count.toLocaleString());

// --- Puzzle 2 ---
const min = Math.min.apply(null, Object.keys(map).map(x => Number(x)));
console.log("2. Number of different ways: ", map[min].toLocaleString());
